RavenDB-17760 Prepare storage to support sharding : introduce a new type of voron index

aviv86 opened the pull request:
### Issue link

https://issues.hibernatingrhinos.com/issue/RavenDB-17760

### Additional description

- Introduce a new type of voron index - `DynamicKeyIndexDef` - which allows to choose the indexing key based on the value
- we will use this index to index items by their `bucket-id` + `etag` (bucket can be computed from the id column) 
- this will allow us to easily `GetItemsByBucket` (ordered by etag on the bucket level) for re-sharding purposes 
- made some abstraction and refactoring in `Table.cs` and `TableSchema.cs` to avoid code duplication (introduced new `AbstractTreeIndexDef`) 
- made some name changes : `SchemaIndexDef => IndexDef`, `FixedSizeSchemaIndexDef => FixedSizeKeyIndexDef`, `TableIndexType => TreeIndexType`
- marked `TableSchema` as partial class and moved all its index definitions to a separate file

### Type of change

- New feature

### How risky is the change?

- Moderate 

### Backward compatibility

- Not relevant

### Is it platform specific issue?

- No

### Documentation update

- No documentation update is needed 

### Testing 

- Tests have been added that prove the fix is effective or that the feature works
- It has been verified by manual testing

### Is there any existing behavior change of other features due to this change?

- No

### UI work

- No UI work is needed

aviv86 committed:
RavenDB-17760 : introduce new schema index which allows you to choose the indexing key based on the value + refactoring

aviv86 committed:
RavenDB-17760 : tests + cleanup

aviv86 committed:
RavenDB-17760 : refactoring and name changes :
- rename SchemaIndexDef => StaticBTreeIndexDef
- rename FixedSizeSchemaIndexDef => FixedSizeTreeIndexDef
- rename CustomSchemaIndexDef => DynamicBTreeIndexDef
- rename TableIndexType.Custom => TableIndexType.Dynamic
- remove TableIndexType.Default (has the same value as TableIndexType.BTree)

ayende commented on src/Voron/Data/Tables/TableSchema.cs:
```diff
@@ -328,31 +317,169 @@ public void Validate(FixedSizeSchemaIndexDef actual)
+                Debug.Assert(IndexValueGenerator.Method.DeclaringType?.AssemblyQualifiedName != null, 
+                    $"Invalid {nameof(IndexValueGenerator)} '{IndexValueGenerator.Method.Name}'");
+
+                var declaringTypeBytes = Encodings.Utf8.GetBytes(IndexValueGenerator.Method.DeclaringType.AssemblyQualifiedName);
```
That is probably a bad idea, we want to skip the versioning here:

`System.Int32, mscorlib, Version=4.0.0.0, Culture=neutral, PublicKeyToken=b77a5c561934e089`

Because we want to be future proof.

aviv86 replied:
done

ayende commented on src/Voron/Data/Tables/TableSchema.cs:
```diff
@@ -328,31 +317,169 @@ public void Validate(FixedSizeSchemaIndexDef actual)
+                currentPtr = input.Read(4, out currentSize);
+                var declaringType = Encodings.Utf8.GetString(currentPtr, currentSize);
+
+                var type = System.Type.GetType(declaringType);
```
For security reasons, we probably want to do that only on types that has a well known attribute that is meant to be used for that.
Otherwise, we may have a code execution on untrusted binaries.

aviv86 replied:
done

ayende reviewed (commented):
This looks quite good.
Once this is done, we need to sit and optimize the manner in which we specialize the tree types.
We can very likely remove the runtime cost abstraction entirely

arekpalinski commented on src/Voron/Data/Tables/AbstractSchemaIndexDefinition.cs:
```diff
@@ -0,0 +1,38 @@
+    public enum TableIndexType
+    {
+        BTree = 0x01,
+        Dynamic = 0x2
```
In practice we have 3 types of table indexes:

 - regular index which uses b+tree under the covers and indexed values are calculated based on provided 
 - new index type which is also b+tree structure but the values are generated dynamically based on provided function
 - fixed size tree index where we deal with keys that are `longs`
 
 So `TableIndexType` enum doesn't really express that. I was thinking about treating it differently - renaming it to `TableBTreeIndexType` and having there:
 
 
 ```
public unsafe class TableSchema
{
   public enum TreeIndexType
   {
      Default = 0x1,
      DynamicKeyValues = 0x2,
   }
...
```

This way we'll have a dedicated types for indexes build based on b+trees (note that I dropped "b" from the name since we don't actually use it in practice  e.g. we don't have `BTree` class but it's just `Tree`, same with `TreePage` not `BTreePage`).

Note that my proposal is to have this enum nested under `TableSchema` so when used outside this class we'll have `TableSchema.TreeIndexType.Default`  - this way it's obvious that is about a table's index

aviv86 replied:
done

arekpalinski commented on src/Voron/Data/Tables/AbstractSchemaIndexDefinition.cs:
```diff
@@ -0,0 +1,38 @@
+        Dynamic = 0x2
+    }
+
+    public abstract class AbstractBTreeIndexDef
```
Given the above comment that we have no 3 types of indexes I wonder if the base class shouldn't be for all of them and dedicated abstract for indexes based on b+trees. My proposal:


```
public abstract class AbstractTableIndexDef<TSelf> where TSelf : AbstractTableIndexDef<TSelf>
{
	public bool IsGlobal;

	public Slice Name;

	public abstract byte[] Serialize();

	public abstract void Validate(TSelf actual);

	public virtual void Validate()
	{
		if (Name.HasValue == false || SliceComparer.Equals(Slices.Empty, Name))
			throw new ArgumentException("Index name must be non-empty", nameof(Name));
	}
}
```

then we could have `AbstractTreeIndexDef`:

```
public abstract class AbstractTreeIndexDef<TSelf> : AbstractTableIndexDef<TSelf> where TSelf : AbstractTreeIndexDef<TSelf>
        {
            public abstract TreeIndexType Type { get; }

            public abstract ByteStringContext.Scope GetValue(ByteStringContext context, ref TableValueReader value,
                out Slice slice);

            public abstract ByteStringContext.Scope GetValue(ByteStringContext context, TableValueBuilder value,
                out Slice slice);
        }
```

with 2 implementations: `IndexDef` (formerly `SchemaIndexDef`) and `DynamicKeyIndexDef` (for new index type introduced here where keys of indexed entries are generated dynamically - hence "key" included there).


Also then we can have `FixedSizeKeyIndexDef` (formely `FixedSizeSchemaIndexDef`) inheriting from `AbstractTableIndexDef` since it has the same common properties like `IsGlobal`, `Name` and methods like `Serialize` or `Validate`:

```
public class FixedSizeKeyIndexDef : AbstractTableIndexDef<FixedSizeSchemaIndexDef>
{
}
```

arekpalinski replied:
I used `<TSelf> where TSelf : AbstractTableIndexDef<TSelf>` construction to have exact type in this method:

```
 public abstract void Validate(TSelf actual);
```

but maybe it's not necessary and having here abstract `AbstractTableIndexDef` is enough

aviv86 replied:
done

arekpalinski commented on src/Voron/Data/Tables/TableSchema.cs:
```diff
@@ -61,26 +53,25 @@ static TableSchema()
+        public Dictionary<Slice, FixedSizeTreeIndexDef> FixedSizeIndexes => _fixedSizeIndexes;
 
-        public class SchemaIndexDef
+        public class StaticBTreeIndexDef : AbstractBTreeIndexDef
```
I like that the indexes classes are nested inside `TableSchema` so we have usages like `new TableSchema.IndexDef()`. Although we could put them in separate files and use partials. So in `TableSchema.IndexDef.cs` we can have:

```
public partial TableSchema
{
  public class IndexDef
  {
  }
}
```

aviv86 replied:
done

arekpalinski commented on src/Voron/Data/Tables/TableSchema.cs:
```diff
@@ -225,38 +187,65 @@ public void Validate(SchemaIndexDef actual)
                         nameof(actual));
 
-                if (StartIndex != actual.StartIndex)
+                if (Type != actual.Type || actual is not StaticBTreeIndexDef staticBTreeIndexDef)
```
Check my above comment about" `<TSelf> where TSelf : AbstractTableIndexDef<TSelf>` construction ". If you decide to go this way then you won't need to cast it here.

aviv86 replied:
decided not to go with `TSelf`, so that `TableSchema` can have `_indexes` dictionary with values of type `AbstractTreeIndexDef` (without type argument)

arekpalinski commented on src/Voron/Data/Tables/TableSchema.cs:
```diff
@@ -328,31 +317,169 @@ public void Validate(FixedSizeSchemaIndexDef actual)
         }
 
-        public TableSchema CompressValues(FixedSizeSchemaIndexDef etagSource, bool compress)
+        public class DynamicBTreeIndexDef : AbstractBTreeIndexDef
```
As proposed above, maybe use `DynamicKeyIndexDef` name for that.

aviv86 replied:
done

arekpalinski commented on src/Voron/Data/Tables/TableSchema.cs:
```diff
@@ -328,31 +317,169 @@ public void Validate(FixedSizeSchemaIndexDef actual)
+        {
+            public override TableIndexType Type => TableIndexType.Dynamic;
+
+            public delegate ByteStringContext.Scope IndexValueAction(ByteStringContext context, ref TableValueReader value, out Slice slice);
```
You were right, it's more about the key of indexed entries not about the values. So maybe `IndexEntryKeyGenerator` would be more suitable?

aviv86 replied:
done

arekpalinski commented on src/Voron/Data/Tables/TableSchema.cs:
```diff
@@ -328,31 +317,169 @@ public void Validate(FixedSizeSchemaIndexDef actual)
+
+            public IndexValueAction IndexValueGenerator;
+
+            public override ByteStringContext.Scope GetSlice(ByteStringContext context, ref TableValueReader value,
```
Currently `FixedSizeSchemaIndexDef` has `GetValue()`. I think it's more accurate than `GetSlice()` so maybe let's use the same here.

aviv86 replied:
done

arekpalinski commented on test/FastTests/Voron/Tables/DynamicBTreeIndex.cs:
```diff
@@ -0,0 +1,213 @@
+
+namespace FastTests.Voron.Tables
+{
+    public unsafe class DynamicBTreeIndex : TableStorageTest
```
I don't like that the name of the test is the same as newly introduced type. Just use issue number?

aviv86 replied:
changed to issue number

arekpalinski commented on src/Voron/Data/Tables/AbstractSchemaIndexDefinition.cs:
```diff
@@ -0,0 +1,38 @@
+using System;
```
Note that the name of file is different `AbstractSchemaIndexDefinition.cs` than class defined here

aviv86 replied:
fixed

arekpalinski commented on test/FastTests/Voron/Tables/DynamicBTreeIndex.cs:
```diff
@@ -0,0 +1,213 @@
+
+                Assert.True(gotValues);
+            }
+        }
```
Can we have some tests which deal with bigger number of items? let's say 10_000 records

aviv86 replied:
done

aviv86 committed:
RavenDB-17760 : address PR comments
- mark TableSchema as partial class and move schema index definitions to a separate partial TableSchema
- rename TableIndexType enum and its values

aviv86 committed:
RavenDB-17760 : address PR comments :
- introduce another abstraction layer for table indexes + renames
- add more tests + change test name
- introduce IndexEntryKeyGeneratorAttribute and verify that the delegate we're using in DynamicKeyIndexDef is marked with this attribute.
- Serialize/ReadFrom - to be future proof, keep just the DeclaringType.FullName + assembly name (without versioning info)

karmeli87 commented on src/Voron/Data/Tables/IndexDefinitions.cs:
```diff
@@ -0,0 +1,443 @@
+                    Name
+                };
+
+                var methodNameBytes = Encodings.Utf8.GetBytes(GenerateKey.Method.Name);
+                fixed (byte* ptr = methodNameBytes)
+                {
+                    serializer.Add(ptr, methodNameBytes.Length);
+                }
+
+                Debug.Assert(GenerateKey.Method.DeclaringType?.FullName != null && GenerateKey.Method.DeclaringType.Assembly.FullName != null,
+                    $"Invalid {nameof(GenerateKey)} '{GenerateKey.Method.Name}'");
+
+                var assemblyName = new AssemblyName(GenerateKey.Method.DeclaringType.Assembly.FullName);
+                var declaringType = $"{GenerateKey.Method.DeclaringType.FullName}, {assemblyName.Name}";
+                var declaringTypeBytes = Encodings.Utf8.GetBytes(declaringType);
+                fixed (byte* ptr = declaringTypeBytes)
+                {
+                    serializer.Add(ptr, declaringTypeBytes.Length);
+                }
```
Why do we care about this at all?

karmeli87 replied:
Also, does it mean if I change the name of the method we will fail here? but if I will change the body it is okay?

ayende replied:
We care about that, yes. Because we want to ensure that we aren't messing up the index.
In general, we would only modify that using schema changes, but it allows us to make more complex decisions later (such as indexes that we can select their form at creation time).

And the idea is that you can change the implementation if needed (but likely will require schema change).

Mostly, to be honest, to ensure that we aren't trying to use an index with the wrong delegate.

arekpalinski commented on src/Voron/Data/Tables/IndexDefinitions.cs:
```diff
@@ -0,0 +1,443 @@
+
+namespace Voron.Data.Tables
+{
+    public unsafe partial class TableSchema
```
Given that this is a partial class let's rename the file to `TableSchema.IndexDefinitions.cs` to follow the convention that we have.

aviv86 replied:
done

arekpalinski commented on src/Voron/IndexEntryKeyGenerator.cs:
```diff
@@ -0,0 +1,9 @@
+namespace Voron
+{
+    [AttributeUsage(AttributeTargets.Method)]
+    public class IndexEntryKeyGeneratorAttribute : Attribute
```
Wouldn't `src/Voron/Data/Tables` be more suitable location for that file?

aviv86 replied:
done

arekpalinski reviewed (commented):
@aviv86 please update the PR description

arekpalinski commented:
@redknightlois Could you have a look at the changes made here? We're introducing a new storage / table index type definition here so we're able to generate an index entry key dynamically - with the usage of a delegate instead of picking it based on column positions.

In particular some abstraction over the index definitions was made here - please check if we don't do any obvious mistakes that could affect the performance.

redknightlois commented on src/Voron/Data/Tables/AbstractSchemaIndexDefinition.cs:
```diff
@@ -0,0 +1,43 @@
+
+        public Slice Name;
+
+        public abstract ByteStringContext.Scope GetSlice(ByteStringContext context, ref TableValueReader value,
```
These are heavy duty calls because of the number of calls. Going abstract here without ensuring devirtualization happens at the JIT level is very risky. We should ensure that it does not regress insertion/mutation/deletion performance. During replication/sharding it may not be that different as those are usually network dominated operations, you have to time this against local operations.

redknightlois replied:
We can apply some of the techniques we created for Corax to solve this type of conundrum, but I will be working on that further down the line with a rework of the performance edge of Voron itself that we have already scheduled with @ayende to happen during the 6.0 development timeframe or even 6.1; I would hesitate a bit to introduce those techniques to solve this before we do all the required work done to push into that direction.

arekpalinski replied:
We have a plan to do some perf tests - https://issues.hibernatingrhinos.com/issue/RavenDB-18076.

But if you think it'd be better / safer to get rid of abstract here then we might adjust that I think. @aviv86 ?

redknightlois replied:
I would say, do them first... lets ensure that we are not regressing hard before continuing. I am more worried about the implications of what that abstraction could impact of the current performance more than to the future.

arekpalinski replied:
Yes, that's the main point of doing it.

aviv86 replied:
> These are heavy duty calls because of the number of calls. Going abstract here without ensuring devirtualization happens at the JIT level is very risky. We should ensure that it does not regress insertion/mutation/deletion performance. During replication/sharding it may not be that different as those are usually network dominated operations, you have to time this against local operations.

what if instead of inheriting from an abstract class, index definitions will implement a common interface - will it make any difference in this aspect? less risky?

karmeli87 replied:
do we have any existing benchmarks for those?

aviv86 replied:
> do we have any existing benchmarks for those?

yes we have a few benchmarks here : 
https://github.com/ravendb/ravendb/tree/v5.3/bench/Voron.Benchmark/Table

redknightlois replied:
> what if instead of inheriting from an abstract class, index definitions will implement a common interface - will it make any difference in this aspect? less risky?

Not really the problem is that the calls become virtual, and at heavy duty places that essentially kills you.

ayende replied:
@redknightlois I intend to sit with @aviv86 next week and change that to generic struct interface to avoid the virtual dispatch, yes. This is done along the way to drive the impl.

ayende replied:
@aviv86 Can you post the before & after results of those benchmarks?

aviv86 replied:
> @aviv86 Can you post the before & after results of those benchmarks?

https://issues.hibernatingrhinos.com/issue/RavenDB-18076#focus=Comments-67-274182.0-0

aviv86 replied:
made some [changes](https://github.com/ravendb/ravendb/pull/13648/commits/0d6ae3790d1d2896e45e28a339f969d021252f68)
- removed `AbstractTableIndexDef` 
- removed all abstract and virtual methods from `AbstractTreeIndexDef` 
- `AbstractTreeIndexDef` now contains only the common properties (Name, IsGlobal, Type) and is used only in "Seek" methods, e.g. : `SeekForwardFromPrefix(AbstractTreeIndexDef index, ...)`,  `SeekBackwardFrom(AbstractTreeIndexDef  index, ...)`, etc.

aviv86 committed:
RavenDB-17760 : address PR comments

ayende commented on src/Voron/Data/Tables/TableSchema.IndexDefinitions.cs:
```diff
@@ -0,0 +1,443 @@
+            {
+                using (context.Allocate(value.Size, out var buffer))
+                {
+                    value.CopyTo(buffer.Ptr);
```
We need to see if we can optimize this. We are forcing a copy of the value here.
A better option would be to have two delegates, or pass both in.

aviv86 replied:
created a separate ticket for optimization 
https://issues.hibernatingrhinos.com/issue/RavenDB-18105

ayende commented on src/Voron/Data/Tables/TableSchema.IndexDefinitions.cs:
```diff
@@ -0,0 +1,443 @@
+
+                var method = type.GetMethod(methodName, BindingFlags.NonPublic | BindingFlags.Public | BindingFlags.Instance | BindingFlags.Static);
+
+                Debug.Assert(method != null, $"Invalid data, failed to get method-info from type : {type}, method name : {methodName}");
+                Debug.Assert(method.IsStatic, $"Invalid data, {nameof(GenerateKey)} must be a static method. method name : {methodName}");
+                Debug.Assert(method.GetCustomAttribute<IndexEntryKeyGeneratorAttribute>() != null, 
+                    $"Invalid data, {nameof(GenerateKey)} must be marked with custom attribute '{nameof(IndexEntryKeyGeneratorAttribute)}'. method name : {methodName}");
+
```
Those need to be _regular checks_. The reason we have those is to prevent arbitrary code execution, so we need that to run even in release mode

aviv86 replied:
done

ayende commented on src/Voron/Data/Tables/TableSchema.IndexDefinitions.cs:
```diff
@@ -0,0 +1,443 @@
+        {
+            public override TreeIndexType Type => TreeIndexType.DynamicKeyValues;
+
+            public delegate ByteStringContext.Scope IndexEntryKeyGenerator(ByteStringContext context, ref TableValueReader value, out Slice slice);
```
We should probably use a function pointer here, I think.

aviv86 committed:
RavenDB-17760 : strip down abstraction layer
- remove AbstractTableIndexDef abstraction layer
- remove all abstract + virtual methods from AbstractTreeIndexDef and keep just the common properties
- TableSchema.cs - store dynamic key indexes in a separate dictionary + handle dynamic key indexes separately
- Table.cs - seek methods 'SeekForwardFrom', 'SeekBackwardFrom', etc. will still operate on abstract type AbstractTreeIndexDef

aviv86 committed:
RavenDB-17760 : address PR comment - change from debug asserts to regular checks in DynamicKeyIndexDef.ReadFrom()
